libs/evm2wasm/evm2wast.c

```diff
@@ -0,0 +1,677 @@
+		"(if(i32.eq(get_local $jump_dest)(i32.const $%ld))(then (br $%ld))(else %s))",
+		segments[i], i, wasm);
+	}
+	size_t s = strlen(wasm) + 358;
+	wasm = realloc(wasm, s);
+	if(wasm == 0) goto err;
+	snprintf(wasm + strlen(wasm), s,
+	"(block $0 (if (i32.eqz (get_global $init)) (then (set_global $init (i32.const 1)) (br $0)) (else (if (i32.eq (get_global $cb_dest) (i32.const 0)) (then %s) (else $cb_dest get_global $cb_dest (set_global $cb_dest (i32.const 0)) (br_table $0 ${brTable}))))))", wasm);
+	free(brtable);
+	return wasm;
+err:
+	perror("Falied to allocate memory for assemble_segments");
+	return 0;
+}
+
+static char *assemble_segments(size_t *segments, size_t len)
+{
+	char *wasm = build_jump_map(segments, len);
+	char *ret = malloc(len*(8+digits(len)+strlen(wasm)+1) + 122);
+	if(ret == 0) goto err;
+	strcat(ret, "(func $main (export \"main\") (local $jump_dest i32) (set_local $jump_dest (i32.const -1)) (block $done (loop $loop ");
+	unsigned int i;
+	for(i=0; i<len; i++)
+	{
+		sprintf(ret+strlen(ret), "(block $%d %s", i+1, wasm);
+	}
+	free(wasm);
+	free(ret);
+	return ret;
+err:
+	perror("Falied to allocate memory for assemble_segments");
+	return 0;
+}
+
+static char *add_stack_check(char *segment)
+{
+	char *check = malloc(1);
+	char *template = malloc(0);
+	if (stack_high != 0)
+	{
+    template = "(if (i32.gt_s (get_global $sp) (i32.const %d))\
+						 (then (unreachable)))";
+		check = realloc(check, strlen(template)+4);
+		if(!check) goto err;
+		snprintf(check, strlen(template)+4, template, (1023-stack_high)*32);
+	}
+	if (stack_low != 0)
+	{
+    template = "(if (i32.lt_s (get_global $sp) (i32.const %d))\
+			       (then (unreachable)))";
+		check = realloc(check, strlen(template)+4);
+		if(!check) goto err;
+		snprintf(check, strlen(template)+4, template,  (-stack_low*32)-32);
+	}
+	segment = realloc(segment, strlen(check) + strlen(segment)+1);
+	sprintf(segment, "%s%s", check, segment);
+	stack_high = 0;
+	stack_low = 0;
+	stack_delta = 0;
+	return segment;
+err:
+	free(template);
+	free(check);
+	perror("cannot alloc memory for stack check templates ");
+	return 0;
+}
+
+static inline void add_metering(char *wast, char *segment)
+{
+	wast = realloc(wast, strlen(wast)+49+strlen(segment));
+	if(!wast)
+	{
+		perror("alloc error while metering wast segment");
+		return;
+	}
+	sprintf(wast, "(call $useGas (i64.const %ld)) %s", gas_count, segment);
+	segment = "";
+	gas_count = 0;
+}
+
+static inline void end_segment(char *wast, char *segment)
+{
+	segment = realloc(segment, strlen(segment)+2);
+	segment = strcat(segment, ")");
+	segment = add_stack_check(segment);
+	add_metering(wast, segment);
+}
+
+static size_t next_jump_dest(char *evm_code, size_t len, size_t i)
+{
+	for(; i<len; i++)
+	{
+		if(evm_code[i] >= PUSH1 && evm_code[i] <= PUSH32)
+		{
+			i+=evm_code[i]-0x59;
+			break;
+		}
+	}
+	return --i;
+}
+char *build_module(char *wast)
+{
+	int i;
+	char *ret = malloc(819200);
```

Comment:
This probably isn't the best idea, in any case at least leave here a big `// FIXME`` comment.